### PR TITLE
fix: react-menu blur handlers are not called on close

### LINKED ISSUE #24427: [Bug]: Menu doesn't call onBlur for contextualmenuitem when the menu is closed
### Library

React / v8 (@fluentui/react)

### System Info

```shell
Windows 11 w/ Edge
```

### Are you reporting Accessibility issue?

yes

### Reproduction

https://codepen.io/makopch/pen/mdxaNrX

## Actual Behavior
1. Open Menu
2. Keyboard between items (notice Focus & Blur output in console)
3. Press Esc to close menu

Nothing is printed to the console

## Expected Behavior
When the menu is closed, the blur even on the focused contextual menu item is fired.

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

Office Online

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### LINKED ISSUE #24546: [Bug]: react-menu doesn't trigger a blur event on the focused menuitem when the menu is closed
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
-
```

### Are you reporting Accessibility issue?

no

### Reproduction

https://codesandbox.io/s/trusting-cloud-js3y1v

## Actual Behavior
The last-focused menuitem will not receive a blur event when the menu is closed, since it is removed from the DOM before focus is called.

## Expected Behavior
If we move the focus change out of a `useEffect`, the blur event will be triggered before the item is removed from the DOM

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by smhigley**
Fixes #24546

Blur handlers on menu items were not being called on close, since focus updates happened in a `useEffect`. I moved the focus change for the closed state out of the `useEffect`, while keeping the focus change for the open state in the `useEffect`.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                                                                             |      Baseline (minified/GZIP) |                            PR |                                                                                                                                                                                                  Change |
| :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: | ----------------------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `188.818 kB`<br />`51.901 kB` | `188.777 kB`<br />`51.914 kB` | `-41 B` <br />`13 B`  |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.fixture.js'>Menu (including children components)</abbr>                                                          | `115.697 kB`<br />`35.316 kB` | `115.655 kB`<br />`35.326 kB` | `-42 B` <br />`10 B`  |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.Selectable.fixture.js'>Menu (including selectable components)</abbr>                                             | `118.896 kB`<br />`35.806 kB` | `118.854 kB`<br />`35.814 kB` |  `-42 B` <br />`8 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                  |        Size (minified/GZIP) |
| -------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr> | `33.19 kB`<br />`10.921 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/744f8e5ccf138ef8c893929c77a1e85f484415d4'>744f8e5ccf138ef8c893929c77a1e85f484415d4</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/Avatar-mount_base.html">1310</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/Avatar-mount.html">1317</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/Button-mount_base.html">946</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/Button-mount.html">957</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProvider-mount_base.html">1579</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProvider-mount.html">1561</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProviderWithTheme-mount_base.html">630</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProviderWithTheme-mount.html">645</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">589</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">586</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">635</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">632</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/MakeStyles-mount_base.html">1901</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/MakeStyles-mount.html">1937</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/SpinButton-mount_base.html">2539</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24532/perf-test-react-components/SpinButton-mount.html">2506</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by layershifter** [comment]
@smhigley a I see the original issue (#24427) is logged for v8, so I am confused with what we are fixing there 🤔 Can you please log an issue for v9 with a repro first (or point to a proper one)? 🙏

**Comment by smhigley** [comment]
Oh, hah. I was chatting with someone directly. Let me file the v9 issue -- a repro for v9 is here: https://codesandbox.io/s/trusting-cloud-js3y1v

File: packages/react-components/react-menu/src/components/Menu/useMenu.tsx [hunk 1]
```diff
@@ -147,18 +161,30 @@ const useMenuOpenState = (
...
+    let shouldHandleTab = false;
+    let pressedShift = false;
+
     if (data.open && e.type === 'contextmenu') {
       state.setContextTarget(e as React.MouseEvent);
     }
 
+    if (data.keyboard) {
+      shouldHandleKeyboard = true;
+      shouldHandleTab = (e as React.KeyboardEvent).key === 'Tab';
```
Comment (bsunderhus) [inline packages/react-components/react-menu/src/components/Menu/useMenu.tsx:174]: Would be nice to use `Tab` from `@fluentui/keyboard-keys` here

**Review by bsunderhus — APPROVED** [review]
LGTM! Another possibility would be to move the handling of focusing on close on the return of the `useEffect` 🤔

File: packages/react-components/react-menu/src/components/Menu/useMenu.tsx [hunk 2]
```diff
@@ -147,18 +161,30 @@ const useMenuOpenState = (
...
       state.setContextTarget(undefined);
-    }
 
-    if (data.keyboard) {
-      shouldHandleKeyboardRef.current = true;
-      shouldHandleTabRef.current = (e as React.KeyboardEvent).key === 'Tab';
-      pressedShiftRef.current = (e as React.KeyboardEvent).shiftKey;
+      if (shouldHandleKeyboard) {
+        if (shouldHandleTab && !state.isSubmenu) {
+          pressedShift ? focusBeforeMenuTrigger() : focusAfterMenuTrigger();
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/Menu/useMenu.tsx:183]:
```suggestion
          if (pressedShift) {
            focusBeforeMenuTrigger()
          } else {
            focusAfterMenuTrigger();
          }
```

nit

**Comment by layershifter** [comment]
> Oh, hah. I was chatting with someone directly. Let me file the v9 issue -- a repro for v9 is here: https://codesandbox.io/s/trusting-cloud-js3y1v
> 
> (updated the PR description to link to the v9 issue)

Nice, thank you 👍 
Can you please unit/e2e tests to cover these changes and ensure that we will not regress? 🙏

**Review by bsunderhus — APPROVED** [review]

**Review by layershifter — APPROVED** [review]
@smhigley please check comments/sync @ling1726 on https://github.com/microsoft/fluentui/issues/24546#issuecomment-1238190984 before merging 🙏

**Comment by smhigley** [comment]
As an update: I chatted with @ling1726 and we both agreed that the `onBlur` issue shouldn't be the reason to take this approach on its own, especially since Ling already filed an issue on React for not following DOM conventions for firing blur when an element is removed.

There is actually another separate reason to follow the approach in this PR, which is that VoiceOver on iOS sometimes jumps its cursor to a different element based on (x, y) screen position when the focused element is removed; then it subsequently does not follow changes to keyboard focus. I was able to reproduce this bug on the react-menu component after fixing the change-focus-on-click issue, and talked with Ling about looking into other fixes like `useLayoutEffect`.

While `useLayoutEffect` seemed to fix it, the current issue is I can no longer reproduce the bug with `useEffect`, so I can't confirm the fix (it's often an on-and-off issue). I'm going to do some more investigating with more iOS devices before continuing :P

**Comment by micahgodbolt** [comment]
> As an update: I chatted with @ling1726 Lingfan Gao FTE and we both agreed that the `onBlur` issue shouldn't be the reason to take this approach on its own, especially since Ling already filed an issue on React for not following DOM conventions for firing blur when an element is removed.
> 
> There is actually another separate reason to follow the approach in this PR, which is that VoiceOver on iOS sometimes jumps its cursor to a different element based on (x, y) screen position when the focused element is removed; then it subsequently does not follow changes to keyboard focus. I was able to reproduce this bug on the react-menu component after fixing the change-focus-on-click issue, and talked with Ling about looking into other fixes like `useLayoutEffect`.
> 
> While `useLayoutEffect` seemed to fix it, the current issue is I can no longer reproduce the bug with `useEffect`, so I can't confirm the fix (it's often an on-and-off issue). I'm going to do some more investigating with more iOS devices before continuing :P

as per the above comment, i'll close this PR down for now. Feel free to re-open if you want to continue this work.

**Pull request closed without merging**